src/server/common/server_base.cpp

```diff
@@ -261,11 +306,41 @@ void server_base::coro_send_doc(socket_ptr socket, simple_wml::document& doc, bo
 		throw;
 	}
 }
+template void server_base::coro_send_doc<socket_ptr>(socket_ptr socket, simple_wml::document& doc, boost::asio::yield_context yield);
+template void server_base::coro_send_doc<tls_socket_ptr>(tls_socket_ptr socket, simple_wml::document& doc, boost::asio::yield_context yield);
+
+template<class SocketPtr> void coro_send_file_userspace(SocketPtr socket, const std::string& filename, boost::asio::yield_context yield)
+{
+	std::size_t filesize { std::size_t(filesystem::file_size(filename)) };
+	union DataSize
+	{
+		uint32_t size;
+		char buf[4];
+	} data_size {};
+	data_size.size = htonl(filesize);
+
+	async_write(*socket, boost::asio::buffer(data_size.buf), yield);
+
+	auto ifs { filesystem::istream_file(filename) };
+	ifs->seekg(0);
+	while(ifs->good()) {
+		char buf[16384];
+		ifs->read(buf, sizeof(buf));
+		async_write(*socket, boost::asio::buffer(buf, ifs->gcount()), yield);
+	}
+}
 
 #ifdef HAVE_SENDFILE
 
-void server_base::coro_send_file(socket_ptr socket, const std::string& filename, boost::asio::yield_context yield)
+template<class SocketPtr> void server_base::coro_send_file(SocketPtr socket, const std::string& filename, boost::asio::yield_context yield)
 {
+	// We fallback to userspace if using TLS socket because sendfile is not aware of TLS state
+	// TODO: keep in mind possibility of using KTLS instead. This seem to be available only in openssl3 branch for now
+	if constexpr (utils::decayed_is_same<tls_socket_ptr, decltype(socket)>) {
+		coro_send_file_userspace(socket, filename, yield);
+		return;
+	} else {
```

Comment:
Do you need this `else` block? If so, it isn't indented, nor is the `return` needed in the first block. Also wonder if this might not be better as a template specialization rather than a constexpr-if check.

`decltype` isn't needed, you can just use `SocketPtr` directly.